torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -353,10 +363,33 @@ struct MutationRemover {
       new_node->output()->setType(node->output()->type());
       mutated_value->replaceAllUsesAfterNodeWith(node, new_node->output());
       node->output()->replaceAllUsesWith(new_node->output());
+
+      // We rewrite something like:
+      // x = torch.zeros()
+      // x.add_(1)
+      // x.add_(2)
+      // to:
+      // x = torch.zeros()
+      // x0 = x.add(1)
+      // x0.add_(2)
+      // For the remainder of the function, x0 will have the
+      // same aliasing relationships as the original x.
+      // To avoid rebuilding the entire alias db, we can replace
+      // the memory dag element of x with x0.
+      aliasDb_->replaceMemoryLocation(mutated_value, new_node->output());
+
+      // it is an invariant that all mutable types have an element in the memory
+      // dag so we must regive x an alias db element. We have already verified
+      // that the mutated value is a fresh alias with a single use.
+      aliasDb_->giveFreshAlias(mutated_value);
+
+      // We must erase the destroyed node from the AliasDb lists of writes
+      aliasDb_->writeIndex_.erase(node);
       node->destroy();
 
-      // :(  TODO: incremental update ?
-      aliasDb_ = torch::make_unique<AliasDb>(graph_);
+      // now that we have removed a mutating op, the write cache is stale
+      // TODO: don't strictly need to reset write cache, evaluate on models
+      aliasDb_->isWriteCacheStale_ = true;
```

Comment:
i think we can also leave `isWriteCacheStale_` false here.

- the mutated value will have one more write than it actually does, however it is safe to consider unmutated values as mutated.
- for the purposes of this pass, we never directly query if a value has writers. We just attempt to move two nodes together, which isn't affected by a slightly stale write cache. It is only affected by the uses / writes of a value between two nodes, and the written node is removed from the graph.

Reply:
What does `target invalidation` mean ?

A pass that mutates the graph already needs to be aware of how it is invalidating the aliasDb. I don't think it makes sense for the pass to generically maintain aliasDb state outside of its own invariants.

Reply:
I might do it as a follow up but leave it like this for now